Show the use of proper gogo/googleapi types with status.

johanbrandhorst opened the pull request:
This avoids the use of `google.golang.org/genproto` package types (except for the mandatory status type).

awalterschulze commented on server/server.go:
```diff
@@ -31,23 +34,25 @@ func (b *Backend) AddUser(ctx context.Context, user *pbExample.User) (*types.Emp
 		}
-
-		return nil, st.Err()
+		byts, _ := proto.Marshal(b)
```
Remember to check the error.

awalterschulze commented on server/server.go:
```diff
@@ -64,6 +69,29 @@ func (b *Backend) ListUsers(_ *types.Empty, srv pbExample.UserService_ListUsersS
+				},
+			},
+		}
+		byts, _ := proto.Marshal(b)
```
Remember to check the error

awalterschulze commented on server/server.go:
```diff
@@ -5,8 +5,11 @@ import (
+	"github.com/gogo/protobuf/proto"
 	"github.com/gogo/protobuf/types"
-	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	pbAny "github.com/golang/protobuf/ptypes/any"
```
Are we still planning to get rid of these pbAny and pbStatus types and replace them with gogo types in future?

awalterschulze commented on main.go:
```diff
@@ -9,12 +9,8 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
-	gogoproto "github.com/gogo/protobuf/proto"
-	golangproto "github.com/golang/protobuf/proto"
```
nice :)

awalterschulze commented on status/README.md:
```diff
@@ -0,0 +1,53 @@
+The changes applies include changing the use of the
+`golang/protobuf` packages to `gogo/protobuf`, and changing the
+generated files from `google.golang.org/genproto/googleapis` to
+`github.com/goog/googleapis/`.
```
gogo not goog :)

johanbrandhorst replied:
Whoops

johanbrandhorst commented:
I'm going to remove the manual vendor changes and wait with merging this until upstream (https://github.com/grpc/grpc-go/pull/1927) has been merged (or rejected :sob:).

awalterschulze commented:
Sounds like a plan :)

On Sun, 18 Mar 2018 at 16:06 Johan Brandhorst <notifications@github.com>
wrote:

> I'm going to remove the manual vendor changes and wait with merging this
> until upstream (grpc/grpc-go#1927
> <https://github.com/grpc/grpc-go/pull/1927>) has been merged (or rejected
> 😭).
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/gogo/grpc-example/pull/10#issuecomment-374006467>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABvsLUSM8jRNZlsiWaP5YvjWAgC1vhO2ks5tfngLgaJpZM4Su5sh>
> .
>

johanbrandhorst commented on Gopkg.toml:
```diff
@@ -45,13 +49,11 @@ required = [
   name = "google.golang.org/grpc"
-  version = "1.10.0"
+# version = "1.10.0"#
+  source = "github.com/johanbrandhorst/grpc-go"
```
TODO: Revert back once https://github.com/grpc/grpc-go/pull/1927 is merged

johanbrandhorst commented on Gopkg.toml:
```diff
@@ -38,20 +38,26 @@ required = [
+  branch = "patch-1"
+  name = "google.golang.org/grpc"
+# version = "1.10.0"#
+  source = "github.com/johanbrandhorst/grpc-go"
```
TODO: Revert this once https://github.com/grpc/grpc-go/pull/1927 is merged

johanbrandhorst commented on Gopkg.toml:
```diff
@@ -38,20 +38,26 @@ required = [
-  name = "github.com/golang/mock"
+  name = "github.com/mwitkow/go-proto-validators"
+
+[[override]]
```
TODO: Set this back to `[[constraint]]`

johanbrandhorst commented:
Updated in anticipation of grpc/grpc-go#1927. Will just need setting the Gopkg.toml back then we can merge.

awalterschulze commented:
Ooo wow.  This looks great.  I hope 1927 goes through.  Great work.

johanbrandhorst committed:
Update dependencies

johanbrandhorst committed:
Switch to using gogo/status for error details.

Adds an error detail to the ListUsers call in order to
ensure both streaming and unary calls properly
translate gogo statuses.

johanbrandhorst committed:
Change gRPC dialing scheme to passthrough.

Upgrading to 1.11.0 revealed that we weren't using the
ipv4 scheme at all as I expected. Now just pass through
the raw localhost address instead.

johanbrandhorst commented:
I've update this with the new gRPC version, and gogo/status, and everything seems to work alright.

awalterschulze reviewed (approved):
Beautiful :)
